lua/cartographer.lua

```diff
@@ -2,20 +2,46 @@
 local api = vim.api
 local Callbacks = require 'cartographer.callbacks'
 
+-- Make a deep copy of opts table
+local function copy_opts(tbl)
+	local new_tbl = {_mode = {}}
+	for key, val in pairs(tbl) do
+		if key ~= '_mode' then new_tbl[key] = val
+		else for k, v in pairs(tbl._mode) do new_tbl._mode[k] = v end
+		end
+	end
+	return new_tbl
+end
+
 --- The tool for building `:map`s. Used as a metatable.
-local MetaCartographer =
+local MetaCartographer
+MetaCartographer =
 {
 	--- Set `key` to `true` if it was not already present
 	--- @param self table the collection of settings
 	--- @param key string the setting to set to `true`
 	--- @returns table self so that this function can be called again
 	__index = function(self, key)
+		local opts = rawget(self, 'opts')
```

Comment:
TBH I added it to avoid using `rawget` :P 

Though with opts separated we could use self to store a cache too.

Reply:
`rawget` does seem to be slower. If we `self = copy_opts(self)` we won't have to use it though, since we haven't `setmetatable` yet. 

I think if we did cache (i.e. `map.x:save()` or some equivalent) we'd definitely add `opts` back for caching.